src/command.go

```diff
@@ -3,84 +3,142 @@ package tome
 )
 
 // Command is a tome Command with its metadata
 type Command struct {
-	Timestamp int64
-	Author    string
-	Tags      []string
-	Command   string
+	Id      uuid.UUID
+	Time    time.Time
+	Author  string
+	Tags    []string
+	Command string
 }
 
-var itemRegex = regexp.MustCompile("(?s)(?P<timestamp>[[:digit:]]+);(?P<author>[^;]+);(?P<tags>(?:[[:word:]]+:?)*);(?P<command>.+)")
+// Helper enum Type for better readability in Deserialize
+type CommandFields int
+
+const (
+	Id = iota
+	Time
+	Author
+	Tags
+	Cmd
+)
 
-func NewCommand(timestamp int64, author string, tags []string, command string) Command {
+func NewCommand(
+	id uuid.UUID,
+	time time.Time,
+	author string,
+	tags []string,
+	command string,
+) Command {
 	return Command{
-		Timestamp: timestamp,
-		Author:    author,
-		Tags:      tags,
-		Command:   command,
+		Id:      id,
+		Time:    time,
+		Author:  author,
+		Tags:    tags,
+		Command: command,
 	}
 }
 
 //String() is the string representation of a Command.
 func (c Command) String() string {
 	t := strings.Join(c.Tags, ":")
-	return fmt.Sprintf(": %d;%s;%s;%s", c.Timestamp, c.Author, t, c.Command)
+	return fmt.Sprintf("%s;%d;%s;%s;%s", c.Id, c.Time.Unix(), c.Author, t, c.EscapedCommandString())
 }
 
-func ParseCommands(reader io.Reader) ([]Command, error) {
+func (c Command) Serialize() string {
+	t := strings.Join(c.Tags, ":")
+	return fmt.Sprintf(": %s;%d;%s;%s;%s", c.Id, c.Time.Unix(), c.Author, t, c.Command)
+}
+
+func (c Command) EscapedCommandString() string {
+	return strings.ReplaceAll(c.Command, "\n", "\\")
+}
+
+func Deserialize(reader io.Reader) ([]Command, error) {
 	var s scanner.Scanner
 	s.Init(reader)
 	s.Whitespace ^= 1<<'\t' | 1<<'\n' | 1<<' '
 
-	lastRune := '\n'
+	lastRune := ' '
+	newCommand := true
+	fieldIndex := 0
+
 	s.IsIdentRune = func(ch rune, i int) bool {
 		oldLastRune := lastRune
 		lastRune = ch
-		return !(ch == ':' && oldLastRune == '\n' || ch == ' ' && i == 0 || ch == scanner.EOF)
+
+		if ch == ':' && oldLastRune == '\n' {
+			newCommand = true
+			return false
+		} else if newCommand && i == 2 {
+			newCommand = false
+			fieldIndex = 0
+			return false
+		} else if ch == ';' && fieldIndex < Cmd {
+			if i == 0 {
+				fieldIndex++
+			}
+			return false
+		} else if ch == scanner.EOF {
+			return false
+		}
+		return true
 	}
 
 	var items []Command
 
+	var id uuid.UUID
+	var timestamp int64
+	var author string
+	var tags []string
+	var command string
+	var lastError error
+
 	for token := s.Scan(); token != scanner.EOF; token = s.Scan() {
-		if token == ':' || token == ' ' {
+		if token == ';' {
 			continue
 		}
 		text := s.TokenText()
-
-		match := itemRegex.FindStringSubmatch(text)
-
-		if match == nil {
-			log.Printf("failed to parse '%s' as Command", text)
+		if text == ": " {
+			lastError = nil
 			continue
 		}
 
-		timestamp, err := strconv.ParseInt(match[itemRegex.SubexpIndex("timestamp")], 10, 64)
-		if err != nil {
-			log.Print(err.Error())
+		if lastError != nil {
```

Comment:
I wanted the parser to be resilient. Once an error occurs it skips all remaining fields in the current 'line'. But it tries again with the next line.

Reply:
The store file should not contain invalid lines, should it?

Reply:
I think tome should never add incompatible lines and if it does (i.e. because of a bug) it should fail when parsing.

Reply:
I disagree. There are several possible reasons: format change, human errors, encoding. In the end it's just a text file in a git repo that everybody is free to change. I see no reason to let the parser crash instead of logging the error and trying to make the best out of the situation.

Reply:
When changing the format we should migrate old entries anyway. It doesn't make sense to defer this and start fresh with 100 errors logged until you do the migration.
What would be human errors? 
By Encoding do you mean format changes? Or actual character encoding?

Reply:
Maybe format change was not a good example.

Actual character encoding  eg. Windows vs. Mac, whatever idk
By human errors I mean someones editing the file manually because he feels like.

Reply:
The point is: What's the benefit of letting the program fail if one line of 10000 lines does not parse?

Reply:
If by editing the file the format is broken it should be fixed. I still think there should be no broken input in this file.

Reply:
Also it would simplify the parser.

Reply:
Ok, I can see that it might make sense especially while developing. Let's merge it like this.